network/network-address/src/lib.rs

```diff
@@ -643,6 +685,124 @@ impl Arbitrary for DnsName {
+/// `&[Protocol]` suffix.
+pub fn parse_memory(protos: &[Protocol]) -> Option<(u16, &[Protocol])> {
+    protos
+        .split_first()
+        .and_then(|(first, suffix)| match first {
+            Protocol::Memory(port) => Some((*port, suffix)),
+            _ => None,
+        })
+}
+
+/// parse the `&[Protocol]` into the `"/ip4/<addr>/tcp/<port>"` or
+/// `"/ip6/<addr>/tcp/<port>"` prefix and unparsed `&[Protocol]` suffix.
+pub fn parse_ip_tcp(protos: &[Protocol]) -> Option<((IpAddr, u16), &[Protocol])> {
+    use Protocol::*;
+
+    if protos.len() < 2 {
+        return None;
+    }
+
+    let (prefix, suffix) = protos.split_at(2);
+    match prefix {
+        [Ip4(ip), Tcp(port)] => Some(((IpAddr::V4(*ip), *port), suffix)),
+        [Ip6(ip), Tcp(port)] => Some(((IpAddr::V6(*ip), *port), suffix)),
+        _ => None,
+    }
+}
+
+/// parse the `&[Protocol]` into the `"/dns/<domain>/tcp/<port>"`,
+/// `"/dns4/<domain>/tcp/<port>"`, or `"/dns6/<domain>/tcp/<port>"` prefix and
+/// unparsed `&[Protocol]` suffix.
+pub fn parse_dns_tcp(protos: &[Protocol]) -> Option<((&DnsName, u16), &[Protocol])> {
+    use Protocol::*;
+
+    if protos.len() < 2 {
+        return None;
+    }
+
+    let (prefix, suffix) = protos.split_at(2);
+    match prefix {
+        [Dns(name), Tcp(port)] => Some(((name, *port), suffix)),
+        [Dns4(name), Tcp(port)] => Some(((name, *port), suffix)),
+        [Dns6(name), Tcp(port)] => Some(((name, *port), suffix)),
+        _ => None,
+    }
+}
+
+/// parse the `&[Protocol]` into the `"/ln-noise-ik/<pubkey>"` prefix and
+/// unparsed `&[Protocol]` suffix.
+pub fn parse_noise_ik(protos: &[Protocol]) -> Option<(&x25519::PublicKey, &[Protocol])> {
+    match protos.split_first() {
+        Some((Protocol::NoiseIK(pubkey), suffix)) => Some((pubkey, suffix)),
+        _ => None,
+    }
+}
+
+/// parse the `&[Protocol]` into the `"/ln-peerid-ex"` prefix and unparsed
+/// `&[Protocol]` suffix.
+pub fn parse_peerid_ex(protos: &[Protocol]) -> Option<&[Protocol]> {
+    match protos.split_first() {
+        Some((Protocol::PeerIdExchange, suffix)) => Some(suffix),
+        _ => None,
+    }
+}
+
+/// parse the `&[Protocol]` into the `"/ln-handshake/<version>"` prefix and
+/// unparsed `&[Protocol]` suffix.
+pub fn parse_handshake(protos: &[Protocol]) -> Option<(u8, &[Protocol])> {
+    match protos.split_first() {
+        Some((Protocol::Handshake(version), suffix)) => Some((*version, suffix)),
+        _ => None,
+    }
+}
+
+/// Extract the second item of a 2-tuple.
+fn snd<T, U>(arg: (T, U)) -> U {
+    arg.1
+}
+
+/// parse canonical libranet protocols
+///
+/// See: `NetworkAddress::is_libranet_addr(&self)`
+fn parse_libranet_protos(protos: &[Protocol]) -> Option<&[Protocol]> {
+    // parse_ip_tcp
+    // <or> parse_dns_tcp
+    // <or> cfg(test) parse_memory
+
+    let opt_transport_suffix = parse_ip_tcp(protos).map(snd);
+    let opt_transport_suffix = opt_transport_suffix.or_else(|| parse_dns_tcp(protos).map(snd));
```

Comment:
```
parse_ip_tcp(protos)
  .or_else(|| parse_dns_tcp(protos)
  .or_else(|| if cfg(test parse_memory(protos) else None)
  .map(snd)?;
```

Reply:
this is about as close as i could get

```rust
    let transport_suffix = None
        .or_else(|| parse_ip_tcp(protos).map(snd))
        .or_else(|| parse_dns_tcp(protos).map(snd))
        .or_else(|| {
            if cfg!(test) {
                parse_memory(protos).map(snd)
            } else {
                None
            }
        })?;
```

not sure which one you prefer

Reply:
to be honest, I don't really care if we compile memory into production, let's not over engineer this.

Reply:
I see the ugliness comes from the `.map(snd)` because the first element is different for every function